### DISCUSSION TITLE
How best to map Flux Kustomization/HelmRelease statuses to a "Synced" status badge?

### DISCUSSION THREAD

**Discussion by nagyv** [fluxcd/flux2#4107] [Q&A] [answered]
At GitLab, we are working on a UI for Flux, and would like to show a synced badge.

Our current proposal is

Status | Conditions
-- | -- 
Synced | at least one { type: Ready, status: "True", reason: ReconciliationSucceeded } |  
Syncing | no synced, at least one { type: Reconciling, status: "True", reason: Progressing\|ProgressingWithRetry } |  
Stalled | no synced, no syncing, at least one { type: Stalled, status: "True" } |  
Failed | no synced, no syncing, no stalled, at least one { type: Ready, status: "False" } |  
Unknown | other combinations |  
Unavailable | no Kustomizations or HelmReleases found |  

This is based on our code reading of Kustomization statuses.

For more details, please see [the GitLab issue](https://gitlab.com/gitlab-org/gitlab/-/issues/419666#note_1491679733)

Do you have any recommendations, ideas to improve the status mapping?

**Comment by stefanprodan** [comment]
I find the `Syncing/Synced` terms confusing, there is no such terminology used in Flux, I think a better term would be `Reconciling/Reconciled`.

**Comment by stefanprodan** [comment, marked as answer]
To determine the reconciling status please see https://github.com/fluxcd/kustomize-controller/blob/main/internal/controller/suite_test.go#L200-L226

**Reply by hiddeco** [reply]
One thing to take into account is that condition reasons are arbitrary and should generally not be used to make any state determinations. Instead, one should resort to condition type and status checking. In combination with e.g. looking at the generation versus observed generation in either the status of the object, or the specific condition type.
